# Assignment 9

Q1. Write a unique paragraph (5-6 sentences) about your favorite topic (e.g., sports,
technology, food, books, etc.).
1. Convert text to lowercase and remove punctuation.
2. Tokenize the text into words and sentences.
3. Remove stopwords (using NLTK's stopwords list).
4. Display word frequency distribuion (excluding stopwords).

In [ ]:
import nltk
import string
import re
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.probability import FreqDist

# # Download necessary NLTK data
# nltk.download('punkt')
# nltk.download('stopwords')

# i
paragraph = """Kyrie Irving is one of the most skilled basketball players in the NBA, known for his incredible handles and mesmerizing finishing ability around the rim.
His signature move, the crossover, has left countless defenders embarrassed on the court, while his clutch shooting while being deffended by Steph Curry helped the Cleveland Cavaliers win their first NBA championship in 2016.
Throughout his career, Irving has demonstrated a unique blend of creativity, athleticism, and technical precision that makes his playing style both effective and aesthetically beautiful to watch.
Despite controversies surrounding his off-court opinions, his basketball talent remains undeniable and continues to inspire a generation of young players around the world.
For the last Ques, I , Sameer , need to add senteces like: don't isn't won't can't.
state-of-the-art technology and cutting-edge
The value of pi is 3.14159. My email is samkh9876@gmail.com.
https://www.example.com. Call us at 123-456-7890 or +91 8130241734.
"""

print("Original Paragraph: ")
print(paragraph)
# i
def clean_text(text):
    text = text.lower()
    text = text.translate(str.maketrans('', '', string.punctuation))
    return text

cleaned_para = clean_text(paragraph)
print(f"Cleaned Text: \n{cleaned_para}")

# ii
sentences = sent_tokenize(paragraph)
words = word_tokenize(cleaned_para)

print(f"Number of sentences: {len(sentences)}")
print("Sentences:")
for i, sentence in enumerate(sentences, 1):
    print(f"{i}. {sentence}")

print(f"\nNumber of words: {len(words)}")
print(f"First 15 tokens: {words[:15]}")

# iii
stop_words = set(stopwords.words('english'))
filtered_words = [word for word in words if word not in stop_words]
print(f"Number of words after stopword removal: {len(filtered_words)}")

# 4. Display word frequency distribution (excluding stopwords)
fdist = FreqDist(filtered_words)
print("Word Frequencies (Top 10):")
for word, frequency in fdist.most_common(10):
    print(f"{word}: {frequency}")


Original Paragraph: 
Kyrie Irving is one of the most skilled basketball players in the NBA, known for his incredible handles and mesmerizing finishing ability around the rim.
His signature move, the crossover, has left countless defenders embarrassed on the court, while his clutch shooting while being deffended by Steph Curry helped the Cleveland Cavaliers win their first NBA championship in 2016.
Throughout his career, Irving has demonstrated a unique blend of creativity, athleticism, and technical precision that makes his playing style both effective and aesthetically beautiful to watch.
Despite controversies surrounding his off-court opinions, his basketball talent remains undeniable and continues to inspire a generation of young players around the world.
For the last Ques, I , ishtpreet, need to add senteces like: don't isn't won't can't.
state-of-the-art technology and cutting-edge
The value of pi is 3.14159. My email is ishpreet.ahuja@gmail.com.
https://www.example.com. Call us a

Q2. Stemming and Lemmatization
1. Take the tokenized words from Question 1 (after stopword removal).
2. Apply stemming using NLTK's PorterStemmer and LancasterStemmer.
3. Apply Lemmatization using NLTK's WordNetLemmatizer.
4. Compare and display results of both techniques.

In [18]:
from nltk.stem import PorterStemmer, LancasterStemmer, WordNetLemmatizer
import pandas as pd

# nltk.download('wordnet')

# Initialize stemmers and lemmatizer
porter_stemmer = PorterStemmer()
lancaster_stemmer = LancasterStemmer()
lemmatizer = WordNetLemmatizer()

# i.
print("First few filtered words:", filtered_words[:10])

# ii.
porter_stemmed = [porter_stemmer.stem(word) for word in filtered_words]
lancaster_stemmed = [lancaster_stemmer.stem(word) for word in filtered_words]

# iii.
lemmatized = [lemmatizer.lemmatize(word) for word in filtered_words]

# iv.
print("\n~~ Comparison ~~")
print(f"{'Original':<15} {'Porter':<15} {'Lancaster':<15} {'Lemmatized':<15}")
print("-" * 60)

for i in range(min(10, len(filtered_words))):  # Limit to first 10 words for simplicity
    print(f"{filtered_words[i]:<15} {porter_stemmed[i]:<15} {lancaster_stemmed[i]:<15} {lemmatized[i]:<15}")

First few filtered words: ['kyrie', 'irving', 'one', 'skilled', 'basketball', 'players', 'nba', 'known', 'incredible', 'handles']

~~ Comparison ~~
Original        Porter          Lancaster       Lemmatized     
------------------------------------------------------------
kyrie           kyri            kyry            kyrie          
irving          irv             irv             irving         
one             one             on              one            
skilled         skill           skil            skilled        
basketball      basketbal       basketbal       basketball     
players         player          play            player         
nba             nba             nba             nba            
known           known           known           known          
incredible      incred          incred          incredible     
handles         handl           handl           handle         


Q3. Regular Expressions and Text Splitting
1. Take their original text from Question 1.
2. Use regular expressions to:
   - a. Extract all words with more than 5 letters.
   - b. Extract all numbers (if any exist in their text).
   - c. Extract all capitalized words.
3. Use text splitting techniques to:
   - a. Split the text into words containing only alphabets (removing digits and special characters).
    - b. Extract words starting with a vowel.

In [19]:
import re

# 1.
print(paragraph)

# 2.
long_words = re.findall(r'\b\w{6,}\b', paragraph)
print(f"Words with more than 5 letters: {long_words}")

numbers = re.findall(r'\d+', paragraph)
print(f"All numbers: {numbers}")

capitalized_words = re.findall(r'\b[A-Z][a-z]*\b', paragraph)
print(f"Capitalized words: {capitalized_words}")

# 3.
alpha_only = re.findall(r'\b[a-zA-Z]+\b', paragraph)
print(f"Words with only alphabets: {alpha_only}")

vowel_words = re.findall(r'\b[aeiouAEIOU]\w*\b', paragraph)
print(f"Words starting with a vowel: {vowel_words}")


Kyrie Irving is one of the most skilled basketball players in the NBA, known for his incredible handles and mesmerizing finishing ability around the rim.
His signature move, the crossover, has left countless defenders embarrassed on the court, while his clutch shooting while being deffended by Steph Curry helped the Cleveland Cavaliers win their first NBA championship in 2016.
Throughout his career, Irving has demonstrated a unique blend of creativity, athleticism, and technical precision that makes his playing style both effective and aesthetically beautiful to watch.
Despite controversies surrounding his off-court opinions, his basketball talent remains undeniable and continues to inspire a generation of young players around the world.
For the last Ques, I , ishtpreet, need to add senteces like: don't isn't won't can't.
state-of-the-art technology and cutting-edge
The value of pi is 3.14159. My email is ishpreet.ahuja@gmail.com.
https://www.example.com. Call us at 123-456-7890 or +91

**Q4. Custom Tokenization & Regex-based Text Cleaning**
1. Take original text from Question 1.
2. Write a custom tokenization function that:
    a. Removes punctuation and special symbols, but keeps contractions (e.g., "isn't" should not be split into "is" and "n't").
    b. Handles hyphenated words as a single token (e.g., "state-of-the-art" remains a single token).
    c. Tokenizes numbers separately but keeps decimal numbers intact (e.g., "3.14" should remain as is).

3. Use Regex Substitutions (re.sub) to:
    a. Replace email addresses with '<EMAIL>' placeholder.
    b. Replace URLs with '<URL>' placeholder.
    c. Replace phone numbers (formats: 123-456-7890 or +91 9876543210) with '<PHONE>' placeholder.


In [20]:
import re

# 1. Use original paragraph
print("Original Paragraph:")
print(paragraph)
print("-" * 40)

# 2. Custom tokenization
def custom_tokenize(text):
    # Replace email, url, phone
    text = re.sub(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b', '<EMAIL>', text)
    text = re.sub(r'https?://[^\s]+', '<URL>', text)
    text = re.sub(r'\b\d{3}-\d{3}-\d{4}\b', '<PHONE>', text)
    text = re.sub(r'\+\d{1,3}\s\d{10}\b', '<PHONE>', text)

    text = re.sub(r"(\w+)\'(\w+)", r"\1_APOS_\2", text)

    hyphens = re.findall(r'\b[\w]+-[\w-]+\b', text)
    for i, word in enumerate(hyphens):
        text = text.replace(word, f"_HYPHEN_{i}_")

    decimals = re.findall(r'\b\d+\.\d+\b', text)
    for i, num in enumerate(decimals):
        text = text.replace(num, f"_DECIMAL_{i}_")
    text = re.sub(r'[^\w\s<>]', ' ', text)

    tokens = text.split()
    tokens = [t.replace('_APOS_', "'") for t in tokens]
    for i, word in enumerate(hyphens):
        tokens = [word if t == f"_HYPHEN_{i}_" else t for t in tokens]

    for i, num in enumerate(decimals):
        tokens = [num if t == f"_DECIMAL_{i}_" else t for t in tokens]
    return tokens, text

tokens, cleaned_text = custom_tokenize(paragraph)

print("Cleaned text with placeholders:")
print(cleaned_text)
print("Tokens:")
print(tokens)


print("\n")
print("Contractions:", [t for t in tokens if "'" in t])
print("Hyphenated words:", [t for t in tokens if "-" in t])
print("Decimals:", [t for t in tokens if re.match(r'\d+\.\d+', t)])
print("Placeholders:", [t for t in tokens if t in ['<EMAIL>', '<URL>', '<PHONE>']])

Original Paragraph:
Kyrie Irving is one of the most skilled basketball players in the NBA, known for his incredible handles and mesmerizing finishing ability around the rim.
His signature move, the crossover, has left countless defenders embarrassed on the court, while his clutch shooting while being deffended by Steph Curry helped the Cleveland Cavaliers win their first NBA championship in 2016.
Throughout his career, Irving has demonstrated a unique blend of creativity, athleticism, and technical precision that makes his playing style both effective and aesthetically beautiful to watch.
Despite controversies surrounding his off-court opinions, his basketball talent remains undeniable and continues to inspire a generation of young players around the world.
For the last Ques, I , ishtpreet, need to add senteces like: don't isn't won't can't.
state-of-the-art technology and cutting-edge
The value of pi is 3.14159. My email is ishpreet.ahuja@gmail.com.
https://www.example.com. Call us at